File: uniqueBinarySearchTreesII.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @return a list of tree node
    def generateTrees(self, n):
    	# @return a list of tree nodes, 
    	# which are root nodes of binary search tree
        def buildTrees(start,end):
        	if start == end:
        		return [TreeNode(start)]
        	if start > end:
        		return [None]
        	roots = []
        	for i in range(start, end+1):
        		leftCandidates = buildTrees(start, i-1)
        		rightCandidates = buildTrees(i+1, end)
        		for left in leftCandidates:
        			for right in rightCandidates:
        				root = TreeNode(i)
        				root.left = left
        				root.right = right
        				roots.append(root)
        	return roots
        return buildTrees(1,n)

File: JAVA/uniqueBinarySearchTreesII.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; left = null; right = null; }
 * }
 */
public class Solution {
    public List<TreeNode> generateTrees(int n) {
        return rangeHelper(1, n);
    }
    
    public List<TreeNode> rangeHelper(int start, int end){
        List<TreeNode> result = new ArrayList<TreeNode>(), left, right;
        if (start > end){
            result.add(null);
            return result;
        }
        if (start == end){
            result.add(new TreeNode(start));
            return result;
        }
        TreeNode root;
        for (int i = start; i <= end; i++){
            left = rangeHelper(start, i-1);
            right = rangeHelper(i+1, end);
            for (TreeNode l:left)
                for (TreeNode r:right){
                    root = new TreeNode(i);
                    root.left = l;
                    root.right = r;
                    result.add(root);
                }
        }
        return result;
    }
}